src/ompl/tools/thunder/SPARSdb.h

```diff
@@ -0,0 +1,802 @@
+*  Redistribution and use in source and binary forms, with or without
+*  modification, are permitted provided that the following conditions
+*  are met:
+*
+*   * Redistributions of source code must retain the above copyright
+*     notice, this list of conditions and the following disclaimer.
+*   * Redistributions in binary form must reproduce the above
+*     copyright notice, this list of conditions and the following
+*     disclaimer in the documentation and/or other materials provided
+*     with the distribution.
+*   * Neither the name of Rutgers University nor the names of its
+*     contributors may be used to endorse or promote products derived
+*     from this software without specific prior written permission.
+*
+*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+*  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+*  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+*  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+/* Author: Andrew Dobson, Dave Coleman */
+
+#ifndef OMPL_TOOLS_THUNDER_SPARS_DB_
+#define OMPL_TOOLS_THUNDER_SPARS_DB_
+
+#include <ompl/geometric/planners/PlannerIncludes.h>
+#include <ompl/datastructures/NearestNeighbors.h>
+#include <ompl/geometric/PathSimplifier.h>
+#include <ompl/util/Time.h>
+
+#include <boost/range/adaptor/map.hpp>
+#include <boost/unordered_map.hpp>
+#include <boost/graph/graph_traits.hpp>
+#include <boost/graph/adjacency_list.hpp>
+#include <boost/graph/filtered_graph.hpp>
+#include <boost/graph/graph_utility.hpp>
+#include <boost/graph/astar_search.hpp>
+#include <boost/graph/connected_components.hpp>
+#include <boost/property_map/property_map.hpp>
+#include <boost/pending/disjoint_sets.hpp>
+#include <boost/function.hpp>
+#include <boost/thread.hpp>
+#include <iostream>
+#include <fstream>
+#include <utility>
+#include <vector>
+#include <map>
+
+namespace ompl
+{
+
+    namespace geometric
+    {
+
+        /**
+           @anchor gSPARSdb
+           @par Short description
+           SPARSdb is a variant of the SPARS algorithm which removes the
+           dependency on having the dense graph, D.  It works through similar
+           mechanics, but uses a different approach to identifying interfaces
+           and computing shortest paths through said interfaces.
+
+           This version has been modified for use with Thunder
```

Comment:
I think this will be very involved to do right, and not worth the extra work (or feasible if I by chance make the next release). SPARS as implemented for OMPL has some differences that will not be easily combined. From the top of my head (its been a while):
- First SPARS builds roadmap from scratch for every problem
- A separate thread waits until the roadmap reaches a certain size/connectivity, then it attempts to plan on this roadmap while the roadmap is still being increased in size
- Once a path is found, SPARS stops growing the roadmap
- There is no easy way to save/load a roadmap from file in the SPARS implementation

In Thunder, I copied the SPARS file then made some modifications and called it SPARSdb. I removed many of the functions that were not necessary to me. I also implemented LazyPRM, so had to modify the graph node data a bit. Still, I am using the same connectivity checks that give us the advantageous theoretical guarantees